### DISCUSSION TITLE
tutorials/intro-typescript/09-codegen

### DISCUSSION THREAD

**Discussion by giscus** [apollographql-education/content-feedback#148] [course feedback]
# tutorials/intro-typescript/09-codegen

https://www.apollographql.com/tutorials/intro-typescript/09-codegen

**Comment by ejaoua** [comment]
the npm run generate produces the following error :  Error: Unable to find Codegen config file! 

        Please make sure that you have a configuration file under the current directory!

**Reply by lizhennessy** [reply]
Hi there @ejaoua ! Can you share your file structure with me? We should have a `codegen.ts` in the root of the project, so outside of the `src` directory.

**Reply by ejaoua** [reply]
sorry for the late reply. Here is the file content : 
```
import type { CodegenConfig } from "@graphql-codegen/cli";

const config: CodegenConfig = {
    schema: "./src/schema.graphql",
    generates: {
        "./src/types.ts": {
            plugins: ["typescript", "typescript-resolvers"],
        },
    },
};

export default config;
```

**Reply by ejaoua** [reply]
I found the fix. in package.json, it should be `"generate": "graphql-codegen --config ./src/codegen.ts"`

**Reply by lizhennessy** [reply]
Glad to hear it @ejaoua ! Let us know if you run into any other problems. 🙂

**Reply by askxue** [reply]
Or, you can directly place codegen.ts in the project root directory, so that graphql-codegen will automatically find this configuration file.

**Comment by DaveNash** [comment]
Hi. 
I'm trying to deploy this to Railway, following [this Side quest](https://www.apollographql.com/tutorials/side-quest-deploy-railway), and my container keeps crashing. I've tried various things to resolve discrepancies between the two code bases; my latest deploy logs indicate that there's no such file when it tries to open /app/dist/schema.graphql. I'm not experienced with much of this, but from comparing the package.json between this course's files and the Side quests, I see a few potential problems: 
1. Railway depends on the 'start' script; here, it calls "npm run compile && nodemon ./dist/index.js"; the Side quest just calls "node src/index". I've updated my 'start' script to call "npm run compile && node ./dist/index.js" Does that seem right? 
2. I've also seen errors indicating it can't find the types module. With src/types.ts included in .gitignore, should "npm run generate" be part of the 'start' script? I tried to include it, but I probably had some syntax wrong. 
3. I have no clue why schema is disappearing on us in .dist. 
Otherwise, great tutorial. Please help if you can; I don't think the Side quest allows for comments.

**Reply by mabuyo** [reply]
Hey there @DaveNash - apologies for the super late reply; this got lost in our comments tracking. I'm sorry we don't currently cover deploying the TypeScript version of the GraphQL API with Railway. We'll make a note of this for future improvements. I would recommend looking at the Railway docs and exploring if they have any TypeScript-specific templates to help you deploy. I found this [TypeScript template](https://railway.app/template/8AWlL5) that could help to look through as an example.

**Comment by ismaelbelghiti** [comment]
Hi, amazing content!
Small typo: codgen => codegen

Best,

**Reply by lizhennessy** [reply]
Thank you @ismaelbelghiti !!! Great catch. We'll fix this.

**Comment by umuterturk** [comment]
`import { Listing } from "../types";` is wrong, 
should be `import { Listing } from "../src/types"; `

**Reply by umuterturk** [reply]
lol I cannot delete my commend :D

**Reply by umuterturk** [reply]
nor edit :D

**Comment by kelvenus** [comment]
I really appreciate that you include how the final code looks at the end of each lessons. 

Being new to coding, I sometimes find it difficult to put the pieces of code together. When you test your code and it fails, your approach gives us a place to correct our work and makes you understand and at least complete the lab. Thanks a million.

**Reply by lizhennessy** [reply]
Love hearing that, @kelvenus ! So glad that this was helpful for you.

**Reply by kelvenus** [reply]
@lizhennessy, the final part of the lab didn't include the final resolvers.ts, or the one provided is incomplete. schema.graphql may also have the same issue. 

When I complete my code, I get an error on Mutation.createListing.

When I complete the basic mutation below, I get the error below. 
 Mutation: {
    createListing: (_, { input }, { dataSources }) => {
      dataSources.listingAPI.createListing(input);
    },
  }
![resolver_ts](https://github.com/user-attachments/assets/afa13d8c-7ca9-4aa6-bf00-775ac220efaf)
![schema_graphql](https://github.com/user-attachments/assets/6e828f01-3353-4c10-af23-c5b75a2990dd)


Type '(_: {}, { input }: RequireFields<MutationCreateListingArgs, "input">, { dataSources }: DataSourceContext) => void' is not assignable to type 'Resolver<ResolverTypeWrapper<Listing>[], {}, DataSourceContext, RequireFields<MutationCreateListingArgs, "input">>'.
  Type '(_: {}, { input }: RequireFields<MutationCreateListingArgs, "input">, { dataSources }: DataSourceContext) => void' is not assignable to type 'ResolverFn<ResolverTypeWrapper<Listing>[], {}, DataSourceContext, RequireFields<MutationCreateListingArgs, "input">>'.
    Type 'void' is not assignable to type 'ResolverTypeWrapper<Listing>[] | Promise<ResolverTypeWrapper<Listing>[]>'.ts(2322)
types.ts(230, 3): The expected type comes from property 'createListing' which is declared here on type 'MutationResolvers<DataSourceContext, {}>'
(property) createListing?: Resolver<ResolverTypeWrapper<Listing>[], {}, DataSourceContext, RequireFields<MutationCreateListingArgs, "input">>

When I complete the final resolvers and schema.graphql, I get this.

src/resolvers.ts(24,69): error TS2345: Argument of type 'CreateListingResponse' is not assignable to parameter of type 'CreateListingInput'.
Type 'CreateListingResponse' is missing the following properties from type 'CreateListingInput': amenities, costPerNight, description, numOfBeds, title
src/resolvers.ts(21,3): error TS1005: ',' expected.
